nspawn: /run/host/ tweaks

poettering opened the pull request:
Let's make some tweaks to /run/host/:

1. Let's place os-release in "flat" style in the dir, in order to not create the illusion we mirror the full FHS hierarchy here (or had plans to). 

2. Let's move the $NOTIFY_SOCKET socket that container payload can use to talk to nspawn to /run/host/socket

3. Let's move the special mount that we use to propagate mounts from host to container (i.e. what is used by "machinectl bind") to /run/host/incoming

Neither 2 nor 3 above fit into the FHS concept, hence let's not copy the paths of os-release 1:1 either.

With this all in place /run/host/ becomes the sole file system location we use to pass data from host to container.

Given we never released the /run/host/ so far, let's change this around, before anyone relies on it.

poettering commented:
/cc @bluca 

I set the v246 milestone, since this touches the /run/host/ API which we should get right before the release

bluca commented:
Thanks!

Paging @smcv - would ```/run/host/usr-lib-os-release``` and ```/run/host/etc-os-release``` work for Flatpak?

bluca commented:
Simon can correct me if I'm wrong, but I think he suggested the ```/run/host/etc/os-release``` (emulate FHS) as Flatpak already supports it for exposing the full host's /etc or /usr tree. In case that option is used, with the flat model, the files would be duplicated. With the current non-flat model, the files are exactly the same and in the same path.
Simon, would the flat model be acceptable?

bluca commented:
For myself and my use case, both implementations work fine, so this change is good for me.

alexlarsson commented:
`/run/host/etc` is where the entire host etc will be if the app has permissions to see it (specifically with host-etc perms, or with generic full filesystem access). This is pre-existing flatpak ABI that can't change at this point. The idea with just mounting that file in the subdir is that the same code would work that already uses these locations.

If systemd wants to use some other form that is fine, but maybe it should use a different toplevel name to not confuse things with the flatpak behavioiur.

mrc0mmand commented on test/units/testsuite-13.sh:
```diff
@@ -66,12 +66,12 @@ if [ -n "${ID:+set}" ] && [ "${ID}" != "${container_host_id}" ]; then exit 1; fi
 '
 
-    systemd-nspawn --register=no -D /testsuite-13.nc-container --bind=/etc/os-release:/tmp/os-release /bin/sh -x -e -c "$_cmd"
+    systemd-nspawn --register=no -D /testsuite-13.nc-container --bind=/etc/os-release:/tmp/etc-os-release --bind=/etc/os-release:/tmp/usr-lib-os-release /bin/sh -x -e -c "$_cmd"
```
```bash
local _cmd='. /tmp/os-release
```
 at the beginning of this function needs to be changes to `/tmp/etc-os-release` as well I guess.

keszybz commented on man/os-release.xml:
```diff
@@ -344,8 +344,8 @@
-    <filename>/run/host/etc/os-release</filename> and
-    <filename>/run/host/usr/lib/os-release</filename>.</para>
+    <filename>/run/host/etc-os-release</filename> and
+    <filename>/run/host/usr-lib-os-release</filename>.</para>
```
I'm not sure if that is an improvement. In particular, still having two files seems pointless. os-release(5) says " Applications should check for the former, and exclusively use its data if it exists". So since we provide `etc-os-release`, the other file should never be used. In that case, I think we should just the the unification ourselves and name the file `/run/host/os-release`.

keszybz commented on src/nspawn/nspawn.c:
```diff
@@ -2517,19 +2517,15 @@ static int setup_propagate(const char *root) {
-                return log_error_errno(r, "Failed to create /run/systemd/nspawn/incoming: %m");
-
-        q = prefix_roota(root, "/run/systemd/nspawn/incoming");
+        q = prefix_roota(root, "/run/host/incoming");
```
I wonder if selinux will barf on this. But let's wait and see.

bluca replied:
it will, at the bare minimum it will need something like:

```
--- a/policy/modules/system/systemd.fc
+++ b/policy/modules/system/systemd.fc
@@ -72,6 +72,7 @@
 /run/systemd/users(/.*)?       gen_context(system_u:object_r:systemd_logind_runtime_t,s0)
 /run/systemd/inhibit(/.*)?     gen_context(system_u:object_r:systemd_logind_inhibit_runtime_t,s0)
 /run/systemd/nspawn(/.*)?      gen_context(system_u:object_r:systemd_nspawn_runtime_t,s0)
+/run/host/incoming(/.*)?       gen_context(system_u:object_r:systemd_nspawn_runtime_t,s0)
 /run/systemd/machines(/.*)?    gen_context(system_u:object_r:systemd_machined_runtime_t,s0)
 /run/systemd/netif(/.*)?       gen_context(system_u:object_r:systemd_networkd_runtime_t,s0)
 

```

poettering commented:
> `/run/host/etc` is where the entire host etc will be if the app has permissions to see it (specifically with host-etc perms, or with generic full filesystem access). 

Hmm, but that's quite a different thing, i.e. I am pretty sure "mounting the whole whole of /etc", and "mounting just os-release" should be two different permissions (i mean, I am not suggesting that flatpak implements things that way, I am just saying philosophically, one being a lot more invasive than the other), and because of that probably should be distinguishable from the payload: it should be ablet to determine if it got the "all-of-etc" perm or the "just-osrelease" perm, but if we'd expose both as /run/host/etc/ then the case "just-osrelease" and the case "all-of-etc, but /etc happens to be empty on the host besides osrelease" would look the same.

> This is pre-existing flatpak ABI that can't change at this point. The idea with just mounting that file in the subdir is that the same code would work that already uses these locations.

I think this doesn't conflict philosophically: i think if sandboxes want to make the whole of /etc available, that's fine, and /run/host/etc/ is a great place for that, but if they want to make just osrelease available then they should do that under a separate name.

> If systemd wants to use some other form that is fine, but maybe it should use a different toplevel name to not confuse things with the flatpak behavioiur.

So let's say flatpak one day adds a permission "full-os-hierarchy", i.e. that the whole OS hierarchy becomes visible to the container. Would you make this appear as top-level /run/host/ mount? Or would you make it available as /run/host/osroot/ or so? i.e. how do you intend /run/host/ to related to the host rootfs?  Do you want this to be a 1:1 mirror of the host tree, but possibly with subtrees pruned off? Or do you intend to use it as a place where various different host provided things can be placed, not just subtree of the host OS tree, but other constructs needed for host-container communication?

With this PR nspawn organizes things like this:

1. /run/host/etc-os-release + /run/host/usr-lib-os-release → the two os-release files
2. /run/host/notify → an sd_notify() compatible socket that PID 1 in the container can use to propagate ready notification to nspawn
3. /run/host/incoming → a directory that has MS_SLAVE mount propagation from a specific host directory enabled and can be used by tools to propagate mounts from the host into the container after the container is up (by mount things into it from the host, and then joining the container's mount namespace and moving it to the final place with MS_MOVE)

[And I intend to use it for other stuff as well later on, for example for a new "credentials" logic, i.e. how certain (usually secret) blobs of data can be passed into system services from the service manager, and into containers from container managers.]

Now, if you intend that /run/host/ is the top of a mirror of the host tree (with subtrees cut off), then this would be incompatible with the use I introduce here. If however you just see /run/host/ as a place where various different bits of host info can be placed, without them being a 1:1 mirror of the host fs tree, but under new names and different hierarchial organization then I think our uses should align somewhat (as long as we don't step on each other's toes regarding the names immediately below /run/hosts).

If the latter is the case we should probably find a new place for the stuff I have in mind, i.e. /run/supervisor or so.

alexlarsson commented:
> Now, if you intend that /run/host/ is the top of a mirror of the host tree (with subtrees cut off), then this would be incompatible with the use I introduce here. If however you just see /run/host/ as a place where various different bits of host info can be placed, without them being a 1:1 mirror of the host fs tree, but under new names and different hierarchial organization then I think our uses should align somewhat (as long as we don't step on each other's toes regarding the names immediately below /run/hosts).
> 
> If the latter is the case we should probably find a new place for the stuff I have in mind, i.e. /run/supervisor or so.

I think we're basically on the same page, but let me explain in more detail how flatpak sees it.

`/run/host` is a directory where we can expose various things from the host. For example we expose the host fonts (wherever the distro configured them to be stored on the host) as `/run/host/fonts`. However, in order to be easily consumed by existing desktop software we chose some directory names such that things work out well if we consider /run/host as a kind of prefix. For example, we always expose host icons as `/run/host/share/icons` and set `XDG_DATA_DIRS` to contain `/run/host/share`, which means icons will be loaded from there.

By default no other host stuff is visible except fonts, icons, tzdata and timezone+resolve config in the sandbox `/etc`. $HOME is available in the regular location, but is a tmpfs, it only has a per-app subdir of $HOME/.var/app/ mounted read-write. 

However, when the app has `--filesystem=host` permissions, the other directories in `/` will be bind-mounted into the sandbox in their regular location, except `/etc`, `/usr`, and `/run` because these are sandbox specific (comes from the runtime the app chose). However with this permission we *do* expose the host `/etc` as `/run/host/etc` and `/usr` as `/run/host/usr` in case the app needs something from there.

Recently we added a subset permission `host-etc` and `host-usr` which *only* does the later (and mount them read-only). These are useful for apps that need to do some custom hacks, but we don't want to expose non-os files that could contain sensitive data.

We also have an open PR (https://github.com/flatpak/flatpak/pull/3733) that adds *just* the `/run/host/etc/os-release` and `/run/host/usr/lib/os-release` files. Whether this should be on by default or require some other type of permission is under discussion.

alexlarsson commented:
Oh, and the PR to just expose the os-release files will place these files in the same place that we would otherwise place them if the app had permissions to see the entire host etc/usr, because we don't want apps to have to look in different places depending on the exact permissions (which are user overridable).

poettering commented:
@alexlarsson hmm, @keszybz is right, it's a bit weird mounting both os-release files individually. Maybe we can agree on simply using the /etc/os-release file if it exists and fall back to /usr/lib/os-release only if it doesn't, and then place it in `/run/host/os-release`.

That way the search path logic for it would be implemented in the container manager, and payload can actually rely on it to be at the same place everywhere.

Would that be agreeable?

(Hmm, and I guess we should mount the whole of /run/host read-only in nspawn, which we currently don't do)

alexlarsson commented:
@poettering That makes sense I guess. But, do you not ever want to load both and combine info from them?

bluca commented:
Don't think so, they should never differ - I think the /etc/os-release path is mostly for backward compatibility, and in modern distros it's always just a relative symlink to /usr/lib/os-release. It's a vendor image specification after all, and both /etc and /usr come from the same vendor.

keszybz commented:
> @poettering That makes sense I guess. But, do you not ever want to load both and combine info from them?

os-release(5) says that /etc/os-release fully replaces /usr/lib/os-release.

alexlarsson commented:
Then i'm fine with flatpak having a single /run/host/os-release always available, with the idividual ones in /run/host/[etc,usr] if the app has additional permissions.

poettering commented:
Yes, /etc/os-release overrides /usr/lib/os-release, if it exists, totally, as documented and as if the /usr/lib version wouldn't exist at all.

(Fedora makes use of this so that they can have distro flavours where they extend the version in /usr/lib/ with some additional fields with some RPM scriptlet and place it in /etc. Terrible idea if you ask me, but the model allows for this.)

bluca commented:
Sounds perfect - thank you all for the great discussion! I'll do the change in Flatpak and push shortly.

@poettering would you like me to take care of the nspawn side, or will you push here?

smcv commented:
> Fedora makes use of this so that they can have distro flavours where they extend the version in /usr/lib/ with some additional fields with some RPM scriptlet and place it in /etc

FWIW, I've used a similar thing on test machines that multi-boot a whole zoo of OSs, so that I can distinguish between parallel installations of Debian 10 without backports, Debian 10 with backports, Debian testing/unstable, and Debian testing/unstable/experimental. If I didn't provide a modified /etc/os-release, I'd have two boot menu entries for Debian 10 and two entries for testing/unstable, and I wouldn't know which one was the one with backports or experimental.

As something the distro does, I'm not 100% convinced by this, but as something a local sysadmin can do, I think it does make sense and is what /etc is for.

poettering commented:
@bluca i will look at it

bluca commented:
> @bluca i will look at it

Ah sorry had just pushed it as https://github.com/systemd/systemd/pull/16563 - will close it then

poettering commented:
> Ah sorry had just pushed it as #16563 - will close it then

Ah, good work. I incorporated most of your commits now in my #16566.

That PR only contains the os-release fixes plus makes /run/hosts read-only to container payload.

The other stuff this PR rearranges let's leave in this PR and do it post release. Dropping from milestone hence.

poettering commented:
I force pushed a new version now, rebased on top of current git. I added docs for the stuff now in /run/host too, and moved some more stuff there.

This should now be ready for review/merge. @bluca maybe something for you to review?

mrc0mmand commented:
`TEST-43-PRIVATEUSER-UNPRIV` failed twice in the row in the same way, it may be related:

```
[   16.789331] testsuite-43.sh[259]: + runas testuser systemd-run --wait --user --unit=test-protect-home-yes -p PrivateUsers=yes -p ProtectHome=yes -P bash -c '
[   16.789331] testsuite-43.sh[259]:         test "$(stat -c %a /home)" = "0"
[   16.789331] testsuite-43.sh[259]:         test "$(stat -c %a /root)" = "0"
[   16.789331] testsuite-43.sh[259]:         test "$(stat -c %a /run/user)" = "0"
[   16.790123] testsuite-43.sh[259]:     '
[   16.790540] testsuite-43.sh[259]: + declare userid=testuser
[   16.791029] testsuite-43.sh[259]: + shift
[   16.791613] testsuite-43.sh[259]: + su testuser -s /bin/sh -c 'XDG_RUNTIME_DIR=/run/user/$UID exec "$@"' -- sh systemd-run --wait --user --unit=test-protect-home-yes -p PrivateUsers=yes -p ProtectHome=yes -P bash -c '
[   16.791613] testsuite-43.sh[259]:         test "$(stat -c %a /home)" = "0"
[   16.791613] testsuite-43.sh[259]:         test "$(stat -c %a /root)" = "0"
[   16.791613] testsuite-43.sh[259]:         test "$(stat -c %a /run/user)" = "0"
[   16.792233] testsuite-43.sh[259]:     '
[   16.819799] audit[291]: USER_AUTH pid=291 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:authentication grantors=pam_rootok acct="testuser" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
[   16.820139] audit[291]: USER_ACCT pid=291 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:accounting grantors=pam_unix acct="testuser" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
[   16.820556] su[291]: (to testuser) root on none
[   16.835419] systemd[1]: n/a: New incoming connection.
[   16.835657] systemd[1]: n/a: Connections of user 0: 0 (of 1024 max)
[   16.835793] systemd[1]: varlink-57: varlink: setting state idle-server
[   16.835923] systemd[1]: varlink-57: New incoming message: {"method":"io.systemd.UserDatabase.GetMemberships","parameters":{"userName":"testuser","service":"io.systemd.DynamicUser"},"more":true}
[   16.836054] systemd[1]: varlink-57: varlink: changing state idle-server → processing-method-more
[   16.836284] systemd[1]: varlink-57: Sending message: {"error":"io.systemd.UserDatabase.NoRecordFound","parameters":{}}
[   16.836436] systemd[1]: varlink-57: varlink: changing state processing-method-more → processed-method
[   16.836567] systemd[1]: varlink-57: varlink: changing state processed-method → idle-server
[   16.837125] audit[291]: CRED_ACQ pid=291 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:setcred grantors=pam_rootok acct="testuser" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
[   16.837289] systemd[1]: varlink-57: Got POLLHUP from socket.
[   16.837866] audit[291]: USER_START pid=291 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:session_open grantors=pam_unix acct="testuser" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
[   16.838248] su[291]: pam_unix(su:session): session opened for user testuser by (uid=0)
[   16.838455] systemd[1]: varlink-57: varlink: changing state idle-server → pending-disconnect
[   16.838638] systemd[1]: varlink-57: varlink: changing state pending-disconnect → processing-disconnect
[   16.838815] systemd[1]: varlink-57: varlink: changing state processing-disconnect → disconnected
[   16.915309] testsuite-43.sh[292]: Running as unit: test-protect-home-yes.service
[   16.943384] testsuite-43.sh[293]: Failed to mount /run/user/4711/systemd/inaccessible/dir to /run/systemd/unit-root/home: No such file or directory
[   16.943384] testsuite-43.sh[293]: test-protect-home-yes.service: Failed to set up mount namespacing: /run/systemd/unit-root/home: No such file or directory
[   16.943977] systemd[293]: test-protect-home-yes.service: Failed at step NAMESPACE spawning /usr/bin/bash: No such file or directory
[   16.945946] systemd[251]: test-protect-home-yes.service: Main process exited, code=exited, status=226/NAMESPACE
[   16.946353] systemd[251]: test-protect-home-yes.service: Failed with result 'exit-code'.
[   16.950368] testsuite-43.sh[292]: Finished with result: exit-code
[   16.950739] testsuite-43.sh[292]: Main processes terminated with: code=exited/status=226
[   16.951164] testsuite-43.sh[292]: Service runtime: 24ms
[   16.951445] testsuite-43.sh[292]: CPU time consumed: 28ms
[   16.956566] su[291]: pam_unix(su:session): session closed for user testuser
[   16.956897] audit[291]: USER_END pid=291 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:session_close grantors=pam_unix acct="testuser" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
[   16.957152] audit[291]: CRED_DISP pid=291 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:setcred grantors=pam_rootok acct="testuser" exe="/usr/bin/su" hostname=? addr=? terminal=? res=success'
[   16.968347] systemd[1]: Received SIGCHLD from PID 259 (bash).
[   16.968752] systemd[1]: Child 259 (bash) died (code=exited, status=226/NAMESPACE)
[   16.968866] systemd[1]: testsuite-43.service: Child 259 belongs to testsuite-43.service.
[   16.969014] systemd[1]: testsuite-43.service: Main process exited, code=exited, status=226/NAMESPACE
[   16.969446] systemd[1]: testsuite-43.service: Failed with result 'exit-code'.
[   16.969767] systemd[1]: testsuite-43.service: Service will not restart (restart setting)
[   16.969911] systemd[1]: testsuite-43.service: Changed start -> failed
[   16.970046] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=64 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[   16.970178] systemd[1]: testsuite-43.service: Job 159 testsuite-43.service/start finished, result=failed
[   16.970283] systemd[1]: Failed to start testsuite-43.service.
```

First full journal: https://ci.centos.org/job/systemd-pr-build-vagrant/8994/artifact//systemd-centos-ci/artifacts_ICUsZ6/vagrant-logs.Cbn/vagrant-arch-testsuite.WTK/TEST-43-PRIVATEUSER-UNPRIV/system.journal
Second full journal: https://ci.centos.org/job/systemd-pr-build-vagrant/8995/artifact//systemd-centos-ci/artifacts_N03XRZ/vagrant-logs.ebC/vagrant-arch-testsuite.OBP/TEST-43-PRIVATEUSER-UNPRIV/system.journal

bluca commented on src/core/mount-setup.c:
```diff
@@ -536,8 +536,17 @@ int mount_setup(bool loaded_policy, bool leave_propagation) {
-        (void) make_inaccessible_nodes(NULL, UID_INVALID, GID_INVALID);
+         * inaccessible nodes from. If we run in a container the host might have created these for us already
+         * in /run/host/inaccessible/. Use those if we can, since tht way we likely get access to block/char
+         * device nodes that are inaccessible, and if userns is used to nodes that ar on mounts owned by a
```
s/ar/are/

bluca commented on docs/CONTAINER_INTERFACE.md:
```diff
@@ -189,6 +196,62 @@ manager, please consider supporting the following interfaces.
+   character and block devices nodes (there all fallbacks for this case).
+
+3. The `/run/host/notify` path is a good choice to place the `sd_notify()`
+   socket in, that may be used for the container PID 1 to report to the container
```
nit: s/container PID 1/container's PID 1/

poettering commented:
Force pushed a new version. Only changes are the ones @bluca  found. And I added one missing "mkdir()", that hopefully fixes the CI failure. Taking liberty to set green label, given @bluca already pretty much blessed this.

bluca commented:
Still failing, but quite strange:


> Aug 18 14:04:43 systemd-testsuite testsuite-43.sh[255]: + runas testuser systemd-run --wait --user --unit=test-private-users -p PrivateUsers=yes -P echo hello
> Aug 18 14:04:43 systemd-testsuite testsuite-43.sh[255]: + declare userid=testuser
> Aug 18 14:04:43 systemd-testsuite testsuite-43.sh[255]: + shift
> Aug 18 14:04:43 systemd-testsuite testsuite-43.sh[255]: + su testuser -s /bin/sh -c 'XDG_RUNTIME_DIR=/run/user/$UID exec "$@"' -- sh systemd-run --wait --user --unit=test-private-users -p PrivateUsers=yes -P echo hello
> ...
> Aug 18 14:04:43 systemd-testsuite su[257]: (to testuser) root on none
> Aug 18 14:04:43 systemd-testsuite systemd[1]: n/a: New incoming connection.
> Aug 18 14:04:43 systemd-testsuite systemd[1]: n/a: Connections of user 0: 0 (of 1024 max)
> Aug 18 14:04:43 systemd-testsuite systemd[1]: varlink-54: varlink: setting state idle-server
> Aug 18 14:04:43 systemd-testsuite systemd[1]: varlink-54: New incoming message: {"method":"io.systemd.UserDatabase.GetMemberships","parameters":{"userName":"testuser","service":"io.systemd.DynamicUser"},"more":true}
> Aug 18 14:04:43 systemd-testsuite systemd[1]: varlink-54: varlink: changing state idle-server → processing-method-more
> **Aug 18 14:04:43 systemd-testsuite systemd[1]: varlink-54: Sending message: {"error":"io.systemd.UserDatabase.NoRecordFound","parameters":{}}**
> ...
> Aug 18 14:04:44 systemd-testsuite testsuite-43.sh[258]: Failed to create bus connection: No such file or directory

Not sure I grok how this patchset could affect that?

poettering commented:
I added another commit now, that hopefully fixes the CI issue. It also hardens generation of the inaccessible inodes a bit. Dropping green label again, since that changes is not trivial. 

ptal!

bluca commented:
Had another look, LGTM!

mrc0mmand commented:
Setting back the green label based on the reviews. Also, I retriggered the failing Semaphore CI, as the fail was a flake.
